serialize.go

```diff
@@ -9,75 +9,137 @@ import (
+	if err != io.EOF {
+		if err := f.Truncate(0); err != nil {
+			j.unlock()
+			return err
+		}
+		mergeEntries(oldEntries, j.entries)
+		j.entries = oldEntries
+	}
+	err = writeJSON(f, j.entries)
+	if errClose := f.Close(); err == nil {
+		err = errClose
+	}
+	if errUnlock := j.unlock(); err == nil {
+		err = errUnlock
+	}
+	return err
 }
 
-// Load uses j.ReadFrom to read cookies
-// from the file at the given path. If the file does not exist,
-// no error will be returned and no cookies
-// will be loaded.
-//
-// The path will be stored in the jar and
-// used when j.Save is next called.
+// TODO:XXX: Redoc
 func (j *Jar) Load(path string) error {
+	j.mu.Lock()
+	defer j.mu.Unlock()
 	j.filename = path
+	if err := j.lock(); err != nil {
+		return err
+	}
 	f, err := os.Open(path)
 	if err != nil {
+		errUnlock := j.unlock()
 		if os.IsNotExist(err) {
-			return nil
+			err = errUnlock
 		}
 		return err
 	}
-	defer f.Close()
-	return j.ReadFrom(f)
+	err = readJSON(f, j.entries)
+	if errClose := f.Close(); err == nil {
+		err = errClose
+	}
+	if errUnlock := j.unlock(); err == nil {
+		err = errUnlock
+	}
+	return err
 }
 
 // WriteTo writes all the cookies in the jar to w
 // as a JSON object.
 func (j *Jar) WriteTo(w io.Writer) error {
-	// TODO don't store non-persistent cookies.
-	encoder := json.NewEncoder(w)
 	j.mu.Lock()
 	defer j.mu.Unlock()
-	if err := encoder.Encode(j.entries); err != nil {
-		return err
-	}
-	return nil
+	return writeJSON(w, j.entries)
+}
+
+// TODO:XXX: Doc
+func writeJSON(w io.Writer, m map[string]map[string]entry) error {
+	// TODO don't store non-persistent cookies.
+	encoder := json.NewEncoder(w)
+	return encoder.Encode(m)
 }
 
 // ReadFrom reads all the cookies from r
 // and stores them in the Jar.
 func (j *Jar) ReadFrom(r io.Reader) error {
-	// TODO verification and expiry on read.
-	decoder := json.NewDecoder(r)
 	j.mu.Lock()
 	defer j.mu.Unlock()
-	if err := decoder.Decode(&j.entries); err != nil {
+	return readJSON(r, j.entries)
+}
+
+// TODO:XXX: Doc
+func readJSON(r io.Reader, m map[string]map[string]entry) error {
+	// TODO verification and expiry on read.
+	decoder := json.NewDecoder(r)
+	return decoder.Decode(&m)
+}
+
+// TODO:XXX: Doc
+func (j *Jar) lock() error {
+	var err error
+	// Spin until lock is acquired.
+	for {
+		j.flock, err = os.OpenFile(j.filename+".lock", os.O_RDWR|os.O_CREATE|os.O_EXCL, 0666)
+		if err == os.ErrExist {
+			time.Sleep(1 * time.Microsecond)
+			continue
+		}
 		return err
 	}
-	return nil
+}
+
+// TODO:XXX: Doc
+func (j *Jar) unlock() error {
+	name := j.flock.Name()
+	err := j.flock.Close()
+	if errRemove := os.Remove(name); err == nil {
+		err = errRemove
+	}
+	return err
+}
+
+// TODO:XXX: Doc
+func mergeEntries(dest, src map[string]map[string]entry) {
+	for k0 := range src {
+		if _, ok := src[k0]; !ok {
+			dest[k0] = make(map[string]entry)
+		}
+		for k1 := range src[k0] {
+			dest[k0][k1] = src[k0][k1]
+		}
```

Comment:
The merging strategy used here seems a little too simplistic. If an existing cookie is updated by a concurrent process then the new version will be replaced with the old one that this current process has in memory. I think we need to merge entries by looking at some kind of last-update time.